### ISSUE TITLE
Separate project and dependency sources for a target

### ISSUE THREAD

**Issue by jastice** [scalacenter/bsp#21]
Currently `buildTarget/dependencySources` will deliver both actual project sources / source directories and library dependency sources in one result.

However these are semantically different. Project sources are part of the actual compilation. Library sources are only used for browsing definitions or generating documentation. Editing them has no effect on the compilation output.

In the general case, it's not possible to differentiate between project sources and library sources based on file/directory type or location. We are currently using heuristics (directory vs jar), but this is not very robust and doesn't even cover all cases which can already be supported. For instance, in IntelliJ it's possible to attach any directory as library sources, and the editor will be aware that it is not part of the project sources.

to be discussed for bsp 2.0

**Comment by jvican** [comment]
The key part here to understand this argument is that the editor may want to show you a popup whenever you modify a scala file (that is writable and in another project) but that is not part of the build. I learned recently via Justin that apparently you can attach source dirs of other builds and replace them by the source jar and IntelliJ will correctly use the local sources.
  
I agree this is an important improvement for BSP 2.0 and I'm in favor of it. I propose we modify `buildTarget/dependencySources` to be:
  
```scala
trait DependencySourcesItem {
  def target: BuildTargetIdentifier

  /** List of resources containing source files of the target's dependencies.
    * They can be changed and can be source files or directories. */
  def mutableSources: List[Uri]


  /** List of resources containing source files of the target's dependencies.
    * They cannot be changed and can be source files, zip files, jar files or directories. */
  def immutableSources: List[Uri]
}
```
  
The duality `mutable/immutable` could be replaced by `editable/persistent` or something along the lines.

**Comment by jvican** [comment]
I think it would make sense to add a field `sources` in `BuildTarget` given how generic this property is for any target. Then, we can delegate `dependencySources` to fetch sources of project dependencies only.
  
I've been thinking about merging this spec change with my proposal in https://github.com/scalacenter/bsp/issues/14 so that the demand of sources of project dependencies is always done by the client explicitly and lazily. However, I think that we should still keep this entrypoint for two reasons:
  
1. Simplifies the life of the client implementations.
2. Some clients may not support this feature at all.
  
Anyhow, it's still premature to tell. I'll get back to this before 2.0.

**Comment by olafurpg** [comment]
I am personally happy with the spec as it is, bloop should just not include source directories of the project in the response for `dependencySources`. I expect this endpoint to only include sources for external dependencies.

There are already two source-related endpoints in the spec

- `buildTarget/textDocuments`: list sources from a build target
- `textDocument/buildTargets`: list build targets for a text document

It seems neither endpoint is implemented in bloop however. I think it would be interesting to update the response type of `buildTarget/textDocuments` to additionally support source directories

```diff
trait BuildTargetTextDocumentsResult {
  /** The source files used by this target */
  def textDocuments: List[TextDocumentIdentifier]
+  /** The directories containing source files for this target */
+  def sourceDirectories: List[URI]
}
```

> I think it would make sense to add a field sources in BuildTarget given how generic this property is for any target.

This conflicts with the idea that `BuildTarget` should be cheap to compute so that `workspace/buildTargets` responds in reasonable time, listing directories for every build target may be expensive for large workspaces.

>  I've been thinking about merging this spec change with my proposal in #14 so the demand of sources of project dependencies is always done by the client explicitly and lazily

I think it's better to have separate endpoints for dependency sources on-demand vs. eagerly. Lazy fetching of sources requires custom development on the client side (at least in metals, I suspect intellij-scala too).

**Comment by olafurpg** [comment]
We could even go further and borrow sbt conventions for managed vs. unmanaged sources.

```diff
trait BuildTargetTextDocumentsResult {
  /** The source files used by this target */
  def textDocuments: List[TextDocumentIdentifier]
+  /** The directories containing user-edited source files for this target.
+    * Should not include source files that are automatically generated by the build. */
+  def unmanagedSourceDirectories: List[URI]
+  /** The directories containing auto-generated source files for this target.
+    * Should not include sources that are manually edited by the user. */
+  def managedSourceDirectories: List[URI]
}
```

**Comment by jastice** [comment]
Sounds reasonable. Does the concept transfer to other build tools?
Maybe the managed/unmanaged/directory should be flags on a `TextDocument` instead since they need not apply to directories. sbt for instance has `managedSources` / `unmanagedSources` which are files.

**Comment by olafurpg** [comment]
After discussions with @jvican we came with the following proposal

- remove `buildTarget/textDocuments`
- bloop should not include project source directories in `buildTarget/dependencySources`
- add new request endpoint `buildTarget/sources`

```scala
// Request from client to server
trait BuildTargetSourcesParams {
  def targets: List[BuildTargetIdenfitier]
}
// Response from server
trait BuildTargetSourcesResult {
  def items: List[SourcesItem]
}
trait SourcesItem {
  /** Represents a directory if the uri ends with a slash /, otherwise a text document */
  def uri: Uri
  def kind: Int
}
object SourcesKind {
  val default = 1
  val autoGenerated = 2
}
```

I'm open to rename `s/SourcesItem/SourceItem/`, but I went with `SourcesItem` for consistency with `DependencySourcesItem`.

**Comment by olafurpg** [comment]
PS, I updated the comment to include the `target BuildTargetIdentifier` and introduce `SourcesItem` and `SourceItem`, please check the website if you read the comment in email.

**Comment by olafurpg** [comment]
Pending fix in #47

**Comment by olafurpg** [comment]
Fixed in #47
